### PR TITLE
feat: implement mixed-precision eigensolver for CG and Davidson methods

### PR DESCRIPTION
**Pull request by laoba657**
- Add PrecisionMode enum (kDouble/kFloat/kMixed) in precision_mode.h
- Implement diag_mixed_precision() for DiagoCG and DiagoDavid
- Add runtime precision configuration via HSolverPW::set_diago_precision_mode()
- Strategy pattern for precision selection (precision_strategy.h)
- Precision analysis documentation (precision_analysis.h)
- Support parse_precision_mode() and precision_mode_to_string() utilities
- Add comprehensive unit tests with gtest (correctness, performance, edge cases)
- Mixed precision: float iteration + double refinement for accuracy < 1e-6

Refs: #mixed-precision #eigensolver

### Reminder
- [ ] Have you linked an issue with this pull request?
- [ ] Have you added adequate unit tests and/or case tests for your pull request?
- [ ] Have you noticed possible changes of behavior below or in the linked issue?
- [ ] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #...

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
- Example: My changes might affect the performance of the application under certain conditions, and I have tested the impact on various scenarios...

### Any changes of core modules? (ignore if not applicable)
- Example: I have added a new virtual function in the esolver base class in order to ...

### CODE REVIEW

File: source/source_hsolver/test/diago_mixed_precision_benchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,570 @@
+/**
+ * @file diago_mixed_precision_benchmark.cpp
+ * @brief 混合精度特征值求解器的性能基准测试和正确性验证
```
Comment (mohanchen) [inline source/source_hsolver/test/diago_mixed_precision_benchmark.cpp]: I would suggest using English for notes.

File: source/source_hsolver/TEST_REPORT.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
+# 混合精度特征值求解器 — 测试结果报告
```
Comment (mohanchen) [inline source/source_hsolver/TEST_REPORT.md]: I would suggest using English for notes.

File: source/source_hsolver/precision_analysis.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
+#ifndef HSOLVER_PRECISION_ANALYSIS_H
+#define HSOLVER_PRECISION_ANALYSIS_H
+
+/**
```
Comment (mohanchen) [inline source/source_hsolver/precision_analysis.h]: I would suggest using English for notes.

**Comment by mohanchen** [comment]
Good job, but you guys need to resolve conflicts first...

File: source/source_hsolver/test/diago_mixed_precision_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,570 @@
...
+#include "source_hsolver/diago_david.h"
+#include <complex>
+#include <random>
+#include <vector>
+#include <chrono>
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <sstream>
+
```
[the remark this answers is not in the document]
Comment (Cstandardlib) [reply source/source_hsolver/test/diago_mixed_precision_benchmark.cpp]: Better refer to the header inclusion pattern of other tests in the directory and follow the same approach.

File: replace.txt
Comment (Cstandardlib) [inline replace.txt:1]: what's this :exploding_head:

File: INPUT_modified
Comment (Cstandardlib) [inline INPUT_modified:1]: Test Input should be placed in corresponding directories rather than the source code.

File: code_stats.sh
Comment (Cstandardlib) [inline code_stats.sh:1]: Only relevant source code, scripts and output should be included

File: dir_list.txt
Comment (Cstandardlib) [inline dir_list.txt:1]: Only relevant source code, scripts and output should be included

File: source/source_hsolver/TEST_REPORT.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+| `diago_cg.cpp` | ✏️ Modified | +2 | 更新枚举引用 |
+| `hsolver_pw.h` | ✏️ Modified | +8 | 精度配置接口 |
+| `hsolver_pw.cpp` | ✏️ Modified | +4 | 传递 PrecisionMode |
+| `test/diago_mixed_precision_benchmark.cpp` | 🆕 New | 420 | 综合测试套件 |
+| `test/CMakeLists.txt` | ✏️ Modified | +8 | New测试目标 |
+| `test/diago_cg_mixed_test.cpp` | ✏️ Modified | +2 | 更新枚举引用 |
+
+---
+
+## 6. Conclusion
```
Comment (Cstandardlib) [inline source/source_hsolver/TEST_REPORT.md]: Test results can be pasted in the PR description or into attachments

File: source/source_hsolver/precision_analysis.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+ *   - 条件数适中: κ(H) < 1e6 时单精度迭代稳定
+ *   - 非刚性谱: 特征值分布不太密集时效果更好
+ *
+ * 不推荐使用混合精度的情况:
+ *   - 极小矩阵 (dim < 50): 开销大于收益
+ *   - 病态矩阵: 条件数过大导致单精度迭代发散
+ *   - 需要极高精度的场景: 误差要求 < 1e-9
+ */
+
+#endif // HSOLVER_PRECISION_ANALYSIS_H
```
Comment (Cstandardlib) [inline source/source_hsolver/precision_analysis.h]: empty header file with comments only
Maybe it's better moved into other headers/source files as comment blocks

**Review by Cstandardlib — CHANGES_REQUESTED** [review]

File: examples/mynotes.txt
Comment (Cstandardlib) [inline examples/mynotes.txt]: notes are not included in cource code

**Comment by Critsium-xy** [comment]
## Blocking issues

1. Unrelated files committed to the repo. These must be removed before merge:
    - INPUT_modified, Si2_INPUT.txt, replace.txt, dir_list.txt, examples/mynotes.txt, code_stats.sh
    - TEST_REPORT.md (208 lines) — benchmark reports belong in the PR description or docs/, not committed as a top-level markdown file.

These look like leftover scratch files from the author's workstation. They should never have been git add-ed.

2. Three new top-level headers for one feature is over-engineered. precision_mode.h, precision_strategy.h, and precision_analysis.h together carry only ~270 lines, and precision_analysis.h is documentation-as-code. Fold the enum + parser into a single header (or inline into diago_cg.h/diago_david.h), and move the analysis text into a comment block or a docs/ page. The Strategy pattern is unjustified weight for a 3-value enum dispatch — a simple switch inside diag() would do.

3. #ifdef ENABLE_MIXED_PRECISION guards are a smell. If the feature is off by default, much of the new code in diago_cg.cpp / diago_david.cpp becomes dead. If it's on, the macro adds nothing. Either commit to the feature (runtime switch only, no compile guard) or keep it on an experimental branch. Don't ship both.

## Algorithmic concerns to verify

4. "Float iterate + double refine" needs a real correctness argument, not just a tolerance check. A 1e-6 agreement on small (≤128) random matrices does not establish behavior on real DFT-PW workloads, where κ(H) can be very large near band gaps and where Davidson's basis orthogonality is sensitive to fp32 roundoff. Please add:
    - A test on a representative physical system (Si bulk, metallic Al — i.e. an actual ABACUS integration test, not GoogleTest with random Hermitian matrices).
    - SCF convergence comparison (final total energy delta, not just eigenvalue delta on iteration 1).
    - Behavior when the float-precision inner loop fails to converge (does the double refinement compensate, or does the solver report a false success?).

5. Performance claims need substantiation. 1.2×–1.8× speedup on N≤128 GoogleTest matrices is not what users care about — they care about wall-clock SCF time on N ~ 10⁴–10⁵ PW
  basis sizes on GPU. The benchmark file should be deleted from tests/ (it's not a unit test) and the numbers re-run on a realistic problem, or the claims removed.

## Code-quality / repo-hygiene

6. Comments and identifiers must be English (reviewer already raised this) — please audit the whole diff, not just the spots called out.
7. No INPUT-parameter wiring shown. hsolver_pw exposes a setter, but is precision_mode exposed in ABACUS's INPUT system? Without that, the feature is unreachable from user input.
8. Constructor signature changes on DiagoCG / DiagoDavid are ABI-breaking for any out-of-tree caller. Prefer a setter or a default-argument additional parameter.

**Comment by Cstandardlib** [comment]
I'd recommend selecting a concrete physical system (like that in examples/) and using a full SCF calculation as the benchmark baseline. What users actually care about is the end-to-end impact of mixed precision on the entire SCF workflow — not just the per-iteration diagonalization speedup.

1. Unittest-level speedup numbers on N≤128 random matrices aren't predictive of real workloads (plane wave basis size N ~ 10⁴).

2. In particular, it's critical to verify whether mixed-precision diagonalization degrades the convergence quality of each SCF step, which could increase the total number of SCF iterations and offset or even reverse the per-step speedup.

**Comment by Critsium-xy** [comment]
Good job! But you need to make the code exacly work and test it out how much it can accelerate the program.

**Comment by mohanchen** [comment]
Thank you for your contributions. I’ve noticed you put a great deal of effort into revising the code and responding to questions. Your implementation has brought fresh insights to the ABACUS community.
However, as the test coverage is still incomplete, we will close this pull request for the time being. Should we revisit relevant code modifications in the future, we will reopen this PR for further discussion.
Thanks again for your hard work.
